# Configuration de la connexion à la base de données  

In [1]:
import pandas as pd 
import cx_Oracle
import sys
import os

%pylab inline
%load_ext sql

%pylab is deprecated, use %matplotlib inline and import the required libraries.
Populating the interactive namespace from numpy and matplotlib


In [2]:
sys.path.append(os.path.abspath('../../.'))
from dbconnect import utilisateur,connexion,password

# Connexion 

In [3]:
%sql oracle+cx_oracle://$utilisateur:$password@$connexion

<img src="https://raw.githubusercontent.com/rbizoi/IntelligenceEnDonneesDeSante/main/InterrogationSQL/images/612.png" width="800">

# L’opérateur égal à


>> SELECT [ ALL | DISTINCT ] <br>
>>> { [{ NOM_TABLE | ALIAS2 }.]* ,<br><br>
>>> [ [{NOM_TABLE |ALIAS2}.]COLONNE1 [AS] ALIAS1[,...]} 

>> FROM NOM_TABLE ALIAS2

>> WHERE PREDICAT<br>



In [4]:
%%sql 
SELECT * 
FROM CLIENTS 
WHERE PAYS='France'
FETCH NEXT 5 ROWS ONLY

 * oracle+cx_oracle://stagiaire:***@51.91.76.248:15440/coursdb
0 rows affected.


[('DUMON', 'Du monde entier', '67, rue des Cinquante Otages', 'Nantes', '44000', 'France', '02.40.67.88.88', '02.40.67.89.89'),
 ('FOLIG', 'Folies gourmandes', '184, chaussée de Tournai', 'Lille', '59000', 'France', '03.20.16.10.16', '03.20.16.10.17'),
 ('FRANR', 'France restauration', '54, rue Royale', 'Nantes', '44000', 'France', '02.40.32.21.21', '02.40.32.21.20'),
 ('LACOR', "La corne d'abondance", "67, avenue de l'Europe", 'Versailles', '78000', 'France', '01.30.59.84.10', '01.30.59.85.11'),
 ('LAMAI', "La maison d'Asie", '1 rue Alsace-Lorraine', 'Toulouse', '31000', 'France', '05.61.77.61.10', '05.61.77.61.11')]

In [5]:
%%sql 
SELECT NOM, PRENOM, DATE_EMBAUCHE 
FROM EMPLOYES 
WHERE DATE_EMBAUCHE = '03/04/1991'

 * oracle+cx_oracle://stagiaire:***@51.91.76.248:15440/coursdb
0 rows affected.


[('Thimoleon', 'Georges', datetime.datetime(1991, 4, 3, 0, 0)),
 ('Maillard', 'Corinne', datetime.datetime(1991, 4, 3, 0, 0))]

In [6]:
%%sql 
SELECT NOM, PRENOM, DATE_EMBAUCHE 
FROM EMPLOYES 
WHERE DATE_EMBAUCHE = to_date('03/04/1991','dd/mm/yyyy')

 * oracle+cx_oracle://stagiaire:***@51.91.76.248:15440/coursdb
0 rows affected.


[('Thimoleon', 'Georges', datetime.datetime(1991, 4, 3, 0, 0)),
 ('Maillard', 'Corinne', datetime.datetime(1991, 4, 3, 0, 0))]

## IS NULL

In [7]:
%%sql 
SELECT NOM, PRENOM, SALAIRE, COMMISSION 
FROM EMPLOYES 
WHERE COMMISSION IS NULL
FETCH NEXT 5 ROWS ONLY

 * oracle+cx_oracle://stagiaire:***@51.91.76.248:15440/coursdb
0 rows affected.


[('Devie', 'Thérèse', Decimal('1540'), None),
 ('Pouetre', 'Camille-Hélène', Decimal('2000'), None),
 ('Ziliox', 'Francoise', Decimal('1900'), None),
 ('Lampis', 'Gabrielle', Decimal('1300'), None),
 ('Fuller', 'Andrew', Decimal('96000'), None)]

## IS NOT NULL

In [8]:
%%sql 
SELECT NOM, PRENOM, SALAIRE, COMMISSION 
FROM EMPLOYES
WHERE COMMISSION IS NOT NULL
FETCH NEXT 5 ROWS ONLY

 * oracle+cx_oracle://stagiaire:***@51.91.76.248:15440/coursdb
0 rows affected.


[('Berlioz', 'Jacques', Decimal('9400'), Decimal('980')),
 ('Nocella', 'Guy', Decimal('7600'), Decimal('910')),
 ('Herve', 'Didier', Decimal('6700'), Decimal('1170')),
 ('Mangeard', 'Jocelyne', Decimal('8200'), Decimal('190')),
 ('Cazade', 'Anne-Claire', Decimal('7200'), Decimal('550'))]

# Les opérateurs logiques

## Supérieur à

In [9]:
%%sql 
SELECT NOM, PRENOM, SALAIRE FROM EMPLOYES WHERE SALAIRE > 13000

 * oracle+cx_oracle://stagiaire:***@51.91.76.248:15440/coursdb
0 rows affected.


[('Leger', 'Pierre', Decimal('19000')),
 ('Fuller', 'Andrew', Decimal('96000')),
 ('Brasseur', 'Hervé', Decimal('147000')),
 ('Splingart', 'Lydia', Decimal('16000')),
 ('Giroux', 'Jean-Claude', Decimal('150000'))]

In [10]:
%%sql 
SELECT NOM, PRENOM, DATE_EMBAUCHE 
FROM EMPLOYES 
WHERE DATE_EMBAUCHE > '30/06/2003'
FETCH NEXT 5 ROWS ONLY

 * oracle+cx_oracle://stagiaire:***@51.91.76.248:15440/coursdb
0 rows affected.


[('Marielle', 'Michel', datetime.datetime(2003, 7, 30, 0, 0)),
 ('Regner', 'Charles', datetime.datetime(2003, 9, 7, 0, 0)),
 ('Fuller', 'Andrew', datetime.datetime(2003, 8, 9, 0, 0)),
 ('Pagani', 'Hector', datetime.datetime(2003, 7, 2, 0, 0)),
 ('Chaussende', 'Maurice', datetime.datetime(2003, 8, 10, 0, 0))]

In [11]:
%%sql 
SELECT NOM,PRENOM 
FROM EMPLOYES 
WHERE NOM > 'Twardowski' 
ORDER BY NOM

 * oracle+cx_oracle://stagiaire:***@51.91.76.248:15440/coursdb
0 rows affected.


[('Urbaniak', 'Isabelle'),
 ('Valot', 'Alain'),
 ('Viry', 'Yvan'),
 ('Weiss', 'Sylvie'),
 ('Ziliox', 'Francoise'),
 ('Zonca', 'Virginie')]

## Inférieur à

In [12]:
%%sql 
SELECT NOM, PRENOM, DATE_EMBAUCHE 
FROM EMPLOYES 
WHERE DATE_EMBAUCHE < '01/01/1990'

 * oracle+cx_oracle://stagiaire:***@51.91.76.248:15440/coursdb
0 rows affected.


[('Peacock', 'Margaret', datetime.datetime(1989, 4, 16, 0, 0)),
 ('Burst', 'Jean-Yves', datetime.datetime(1988, 10, 23, 0, 0)),
 ('Malejac', 'Yannick', datetime.datetime(1987, 11, 29, 0, 0)),
 ('Maurer', 'Véronique', datetime.datetime(1989, 5, 27, 0, 0)),
 ('Alvarez', 'Marcel', datetime.datetime(1988, 8, 3, 0, 0))]

In [13]:
%%sql 
SELECT NOM_PRODUIT, NO_FOURNISSEUR 
FROM PRODUITS 
WHERE NO_FOURNISSEUR <= 2
FETCH NEXT 5 ROWS ONLY

 * oracle+cx_oracle://stagiaire:***@51.91.76.248:15440/coursdb
0 rows affected.


[('Chai', 1),
 ('Chang', 1),
 ('Aniseed Syrup', 1),
 ("Chef Anton's Cajun Seasoning", 2),
 ('Louisiana Hot Spiced Okra', 2)]

In [14]:
%%sql 
SELECT NOM_PRODUIT, QUANTITE 
FROM PRODUITS 
WHERE QUANTITE >'50'

 * oracle+cx_oracle://stagiaire:***@51.91.76.248:15440/coursdb
0 rows affected.


[('Manjimup Dried Apples', '50 cartons (300 g)'),
 ('Thüringer Rostbratwurst', '50 sacs x 30 saucisses'),
 ('Dried Apples', '50\xa0paquets de 300\xa0g'),
 ('Amandes', 'Paquet de 5\xa0kg'),
 ('Dried Plums', 'Sac de 450\xa0g')]

In [15]:
%sql SELECT ASCII(5), ASCII('P') FROM DUAL

 * oracle+cx_oracle://stagiaire:***@51.91.76.248:15440/coursdb
0 rows affected.


[(53, 80)]

In [16]:
%sql SELECT CODE_CLIENT, SOCIETE FROM CLIENTS WHERE CODE_CLIENT < 'B'

 * oracle+cx_oracle://stagiaire:***@51.91.76.248:15440/coursdb
0 rows affected.


[('ALFKI', 'Alfreds Futterkiste'),
 ('ANATR', 'Ana Trujillo Emparedados y helados'),
 ('ANTON', 'Antonio Moreno Taquería'),
 ('AROUT', 'Around the Horn')]

## Différent de

In [17]:
%%sql 
SELECT NOM, PRENOM, FONCTION 
FROM EMPLOYES 
WHERE FONCTION <>'Représentant(e)'
FETCH NEXT 10 ROWS ONLY

 * oracle+cx_oracle://stagiaire:***@51.91.76.248:15440/coursdb
0 rows affected.


[('Devie', 'Thérèse', 'Assistante commerciale'),
 ('Ragon', 'André', 'Chef des ventes'),
 ('Pouetre', 'Camille-Hélène', 'Assistante commerciale'),
 ('Leger', 'Pierre', 'Chef des ventes'),
 ('Ziliox', 'Francoise', 'Assistante commerciale'),
 ('Lampis', 'Gabrielle', 'Assistante commerciale'),
 ('Belin', 'Chantal', 'Chef des ventes'),
 ('Fuller', 'Andrew', 'Vice-Président'),
 ('Brasseur', 'Hervé', 'Vice-Président'),
 ('Poupard', 'Claudette', 'Assistante commerciale')]

# L’opérateur LIKE
Il utilise deux caractères spéciaux pour signifier le type de correspondance recherchée :<br>
* un signe pourcentage « **%** », appelé caractère générique, 
* et un caractère de soulignement « **_** », appelé marqueur de position.


In [18]:
%%sql SELECT NOM_PRODUIT, QUANTITE 
FROM PRODUITS
WHERE QUANTITE LIKE '%boîtes%kg%'

 * oracle+cx_oracle://stagiaire:***@51.91.76.248:15440/coursdb
0 rows affected.


[('Konbu', '1 boîtes (2 kg)'),
 ('Alice Mutton', '20 boîtes (1 kg)'),
 ('Filo Mix', '16 boîtes (2 kg)'),
 ('Long Grain Rice', '16\xa0boîtes de 2\xa0kg')]

Les valeurs contenues dans les colonnes sont sensibles à la case (majuscule, minuscule), les informations saisies dans les chaînes de caractères de comparaison doivent l’être aussi.

In [19]:
%%sql 
SELECT NOM_PRODUIT,QUANTITE 
FROM PRODUITS 
WHERE QUANTITE LIKE '__0%g'

 * oracle+cx_oracle://stagiaire:***@51.91.76.248:15440/coursdb
0 rows affected.


[('Fruit Cocktail', '430\xa0g'),
 ('Pears', '430\xa0g'),
 ('Peaches', '430\xa0g'),
 ('Pineapple', '430\xa0g'),
 ('Cherry Pie Filling', '430\xa0g'),
 ('Green Beans', '400\xa0g'),
 ('Corn', '400\xa0g'),
 ('Peas', '400\xa0g'),
 ('Tuna Fish', '140\xa0g'),
 ('Smoked Salmon', '140\xa0g')]

In [20]:
%%sql 
SELECT NOM_PRODUIT, QUANTITE 
FROM PRODUITS 
WHERE QUANTITE LIKE '___%pièces' 

 * oracle+cx_oracle://stagiaire:***@51.91.76.248:15440/coursdb
0 rows affected.


[('Teatime Chocolate Biscuits', '10 boîtes x 12 pièces'),
 ("Sir Rodney's Scones", '24 cartons x 4 pièces'),
 ('Perth Pasties', '48 pièces'),
 ('Escargots de Bourgogne', '24 pièces'),
 ('Wimmers gute Semmelknödel', '20 sacs x 4 pièces'),
 ('Scottish Longbreads', '10 sacs x 8 pièces'),
 ('Chocolate Biscuits Mix', '10\xa0boîtes de 12\xa0pièces'),
 ('Scones', '24\xa0paquets de 4\xa0pièces')]

In [21]:
%%sql 
SELECT NOM_CATEGORIE, DESCRIPTION 
FROM CATEGORIES 
WHERE DESCRIPTION LIKE '%' || NOM_CATEGORIE ||'%'

 * oracle+cx_oracle://stagiaire:***@51.91.76.248:15440/coursdb
0 rows affected.


[('Boissons', 'Boissons, cafés, thés, bières'),
 ('Desserts', 'Desserts et friandises'),
 ('Viandes', 'Viandes préparées'),
 ('Viande en conserve', 'Viande en conserve')]

# Les expressions régulières

In [22]:
%%sql 
SELECT QUANTITE 
FROM PRODUITS
WHERE REGEXP_LIKE(QUANTITE,'^(10).*(pièces)$')
FETCH NEXT 10 ROWS ONLY

 * oracle+cx_oracle://stagiaire:***@51.91.76.248:15440/coursdb
0 rows affected.


[('10 boîtes x 12 pièces',),
 ('10 sacs x 8 pièces',),
 ('10\xa0boîtes de 12\xa0pièces',)]

In [23]:
%%sql 
SELECT QUANTITE 
FROM PRODUITS 
WHERE REGEXP_LIKE (QUANTITE,'[,x]')
FETCH NEXT 10 ROWS ONLY

 * oracle+cx_oracle://stagiaire:***@51.91.76.248:15440/coursdb
0 rows affected.


[('10 boîtes x 20 sacs',),
 ('10 boîtes x 12 pièces',),
 ('24 cartons x 4 pièces',),
 ('24 boîtes x 2 tartes',),
 ('50 sacs x 30 saucisses',),
 ('20 sacs x 4 pièces',),
 ('10 sacs x 8 pièces',),
 ('24 bouteilles (0,5 litre)',),
 ('48\xa0bocaux de 170\xa0g',),
 ('12\xa0bocaux de 225\xa0g',)]

In [24]:
%%sql 
SELECT QUANTITE 
FROM PRODUITS 
WHERE REGEXP_LIKE (QUANTITE,'(sacs).*(pièces)')

 * oracle+cx_oracle://stagiaire:***@51.91.76.248:15440/coursdb
0 rows affected.


[('20 sacs x 4 pièces',), ('10 sacs x 8 pièces',)]

In [25]:
%%sql 
SELECT QUANTITE 
FROM PRODUITS 
WHERE REGEXP_LIKE (QUANTITE,'^2.*(carton|pièces|bouteilles)')

 * oracle+cx_oracle://stagiaire:***@51.91.76.248:15440/coursdb
0 rows affected.


[('24 bouteilles (1 litre)',),
 ('24 cartons x 4 pièces',),
 ('24 cartons (500 g)',),
 ('24 cartons (200 g)',),
 ('24 bouteilles (70 cl)',),
 ('24 bouteilles (1 litre)',),
 ('24 cartons (50 g)',),
 ('24 cartons (250 g)',),
 ('24 bouteilles (12 onces)',),
 ('24 bouteilles (355 ml)',),
 ('24 cartons (200 g)',),
 ('24 bouteilles (250 ml)',),
 ('20 cartons (2 kg)',),
 ('24 cartons (250 g)',),
 ('24 pièces',),
 ('24 bouteilles (500 ml)',),
 ('20 sacs x 4 pièces',),
 ('24 bouteilles (0,5 litre)',),
 ('24\xa0paquets de 4\xa0pièces',),
 ('24\xa0bouteilles de 35\xa0cl',)]

In [26]:
%%sql 
SELECT QUANTITE 
FROM PRODUITS 
WHERE REGEXP_LIKE (QUANTITE,'^[^1-4]')

 * oracle+cx_oracle://stagiaire:***@51.91.76.248:15440/coursdb
0 rows affected.


[('50 cartons (300 g)',),
 ('50 sacs x 30 saucisses',),
 ('50\xa0paquets de 300\xa0g',),
 ('Paquet de 5\xa0kg',),
 ('Sac de 450\xa0g',)]

In [27]:
%%sql 
SELECT SOCIETE 
FROM FOURNISSEURS 
WHERE REGEXP_LIKE ( SOCIETE,'(s.r.l)')

 * oracle+cx_oracle://stagiaire:***@51.91.76.248:15440/coursdb
0 rows affected.


[('Formaggi Fortini s.r.l.',), ('Pasta Buttini s.r.l.',)]

In [28]:
%%sql 
SELECT SOCIETE 
FROM FOURNISSEURS 
WHERE REGEXP_LIKE ( SOCIETE,'([AEIOU])1', 'i')

 * oracle+cx_oracle://stagiaire:***@51.91.76.248:15440/coursdb
0 rows affected.


[]

In [29]:
%%sql 
SELECT CODE_POSTAL,SOCIETE, VILLE  
FROM FOURNISSEURS 
WHERE REGEXP_LIKE(CODE_POSTAL,'^[0-9]+$')

 * oracle+cx_oracle://stagiaire:***@51.91.76.248:15440/coursdb
0 rows affected.


[('48104', 'Kelly', 'Ann Arbor'),
 ('70117', 'Nouvelle-Orléans Cajun Delights', 'Nouvelle-Orléans'),
 ('48104', "Grandma Kelly's Homestead", 'Ann Arbor'),
 ('100', 'Tokyo Traders', 'Tokyo'),
 ('33007', "Cooperativa de Quesos 'Las Cabras'", 'Oviedo'),
 ('545', "Mayumi's", 'Osaka'),
 ('3058', 'Pavlova, Ltd.', 'Melbourne'),
 ('5442', 'Refrescos Americanas LTDA', 'São Paulo'),
 ('10785', 'Heli Süßwaren GmbH  Co. KG', 'Berlin'),
 ('60439', 'Plutzer Lebensmittelgroßmärkte AG', 'Frankfurt a.M.'),
 ('27478', 'Nord-Ost-Fisch Handelsgesellschaft mbH', 'Cuxhaven'),
 ('48100', 'Formaggi Fortini s.r.l.', 'Ravenna'),
 ('1320', 'Norske Meierier', 'Sandvika'),
 ('97101', 'Bigfoot Breweries', 'Bend'),
 ('75004', 'Aux joyeux ecclésiastiques', 'Paris'),
 ('02134', 'New England Seafood Cannery', 'Boston'),
 ('0512', 'Leka Trading', 'Singapore'),
 ('2800', 'Lyngbysild', 'Lyngby'),
 ('53120', 'Karkki Oy', 'Lappeenranta'),
 ('2042', "G'day, Mate", 'Sydney'),
 ('84100', 'Pasta Buttini s.r.l.', 'Salerno'),
 ('71300', 'Escargots Nouveaux', 'Montceau'),
 ('74000', 'Gai pâturage', 'Annecy')]

In [30]:
%%sql 
SELECT CODE_POSTAL,SOCIETE, VILLE  
FROM FOURNISSEURS 
WHERE REGEXP_LIKE(CODE_POSTAL,'^d+$')

 * oracle+cx_oracle://stagiaire:***@51.91.76.248:15440/coursdb
0 rows affected.


[]

In [33]:
%%sql 
SELECT CODE_POSTAL,SOCIETE, VILLE  
FROM FOURNISSEURS 
WHERE REGEXP_LIKE(CODE_POSTAL,'^[[:digit:]]+$')

 * oracle+cx_oracle://stagiaire:***@51.91.76.248:15440/coursdb
0 rows affected.


[('48104', 'Kelly', 'Ann Arbor'),
 ('70117', 'Nouvelle-Orléans Cajun Delights', 'Nouvelle-Orléans'),
 ('48104', "Grandma Kelly's Homestead", 'Ann Arbor'),
 ('100', 'Tokyo Traders', 'Tokyo'),
 ('33007', "Cooperativa de Quesos 'Las Cabras'", 'Oviedo'),
 ('545', "Mayumi's", 'Osaka'),
 ('3058', 'Pavlova, Ltd.', 'Melbourne'),
 ('5442', 'Refrescos Americanas LTDA', 'São Paulo'),
 ('10785', 'Heli Süßwaren GmbH  Co. KG', 'Berlin'),
 ('60439', 'Plutzer Lebensmittelgroßmärkte AG', 'Frankfurt a.M.'),
 ('27478', 'Nord-Ost-Fisch Handelsgesellschaft mbH', 'Cuxhaven'),
 ('48100', 'Formaggi Fortini s.r.l.', 'Ravenna'),
 ('1320', 'Norske Meierier', 'Sandvika'),
 ('97101', 'Bigfoot Breweries', 'Bend'),
 ('75004', 'Aux joyeux ecclésiastiques', 'Paris'),
 ('02134', 'New England Seafood Cannery', 'Boston'),
 ('0512', 'Leka Trading', 'Singapore'),
 ('2800', 'Lyngbysild', 'Lyngby'),
 ('53120', 'Karkki Oy', 'Lappeenranta'),
 ('2042', "G'day, Mate", 'Sydney'),
 ('84100', 'Pasta Buttini s.r.l.', 'Salerno'),
 ('71300', 'Escargots Nouveaux', 'Montceau'),
 ('74000', 'Gai pâturage', 'Annecy')]

In [40]:
%%sql 
SELECT SOCIETE, PAYS, TELEPHONE 
FROM CLIENTS  
WHERE REGEXP_LIKE ( TELEPHONE, '^[0-9][0-9](\.[0-9][0-9])+$' )

 * oracle+cx_oracle://stagiaire:***@51.91.76.248:15440/coursdb
0 rows affected.


[('Du monde entier', 'France', '02.40.67.88.88'),
 ('Folies gourmandes', 'France', '03.20.16.10.16'),
 ('France restauration', 'France', '02.40.32.21.21'),
 ("La corne d'abondance", 'France', '01.30.59.84.10'),
 ("La maison d'Asie", 'France', '05.61.77.61.10'),
 ('Paris spécialités', 'France', '01.42.34.22.66'),
 ('Spécialités du monde', 'France', '01.47.55.60.10'),
 ('Victuailles en stock', 'France', '04.78.32.54.86'),
 ('Vins et alcools Chevalier', 'France', '03.26.47.15.10'),
 ('Blondel père et fils', 'France', '03.88.60.15.31'),
 ("Bon app'", 'France', '04.91.24.45.40')]

# Atelier 3.1	Les opérateurs logiques
Écrivez les requêtes vous permettant d’afficher :
>> * Le nom, le prénom et la date de naissance de tous les employés de la société.<br>
>> * Le nom de la société, la ville et le pays de tous les fournisseurs.<br>
>> * La fonction de tous les employés.<br>
>> * Toutes les fonctions des employés de l’entreprise, chaque fonction doit être affichée une seule fois.<br>
>> * La liste des localités dans lesquelles la société a au moins un client.<br>


## La restriction
Écrivez les requêtes permettant d’afficher :
>> * Le nom de la société et de la localité des clients qui habitent à **Toulouse**.<br>
>> * Le nom, le prénom et la fonction des employés qui ne sont pas des représentants.<br>
>> * Le nom du produit, la catégorie et le fournisseur des produits qui ne sont pas disponibles, le champ **INDISPONIBLE** est égal à 1.<br>
>> * Le nom, prénom et fonction des employés qui ont un salaire inférieur à **3500**.<br>
>> * Le nom, prénom et fonction des employés dirigés par l’employé numéro **86**.<br>
>> * Le nom, prénom et fonction des employés recrutés après **01/01/2003**.<br>

## Le traitement des chaînes de caractères
Écrivez les requêtes permettant d’afficher :
>> * Les produits et leur quantité conditionnée en bouteilles d’un litre.<br>

# L’opérateur BETWEEN

In [ ]:
%%sql 
SELECT NOM, PRENOM, SALAIRE 
FROM EMPLOYES 
WHERE SALAIRE BETWEEN 1500 AND 2000

In [ ]:
%%sql 
SELECT NOM, PRENOM, DATE_NAISSANCE 
FROM EMPLOYES 
WHERE DATE_NAISSANCE BETWEEN '01/01/1985' AND '31/12/1985'

In [ ]:
%%sql 
SELECT CODE_CLIENT, SOCIETE, VILLE  
FROM CLIENTS 
WHERE CODE_CLIENT BETWEEN 'GALED' AND 'HANAR'

# L’opérateur IN

In [ ]:
%%sql 
SELECT SOCIETE, VILLE  
FROM CLIENTS 
WHERE VILLE IN ('Paris','Strasbourg','Toulouse')

# Les expressions logiques
## AND

In [ ]:
%%sql 
SELECT NOM_PRODUIT,QUANTITE, UNITES_STOCK 
FROM PRODUITS 
WHERE  UNITES_STOCK > 100 
  AND  QUANTITE LIKE '%boîte%'

In [ ]:
%%sql 
SELECT NO_COMMANDE, CODE_CLIENT, NO_EMPLOYE, DATE_ENVOI 
FROM COMMANDES 
WHERE CODE_CLIENT = 'COMMI' 
  AND NO_EMPLOYE BETWEEN 6 AND 8 
  AND DATE_ENVOI < '01/04/2010'

## OR

In [ ]:
%%sql 
SELECT NOM_PRODUIT, QUANTITE, UNITES_STOCK 
FROM PRODUITS 
WHERE QUANTITE IS NULL OR UNITES_STOCK IS NULL

In [ ]:
%%sql 
SELECT NOM_PRODUIT, CODE_CATEGORIE 
FROM PRODUITS 
WHERE NO_FOURNISSEUR = 1 OR NO_FOURNISSEUR = 5

In [ ]:
%%sql 
SELECT SOCIETE, PAYS, FAX 
FROM FOURNISSEURS 
WHERE FAX IS NULL AND PAYS = 'Allemagne' OR PAYS = 'France'

In [ ]:
%%sql 
SELECT SOCIETE, PAYS, FAX 
FROM FOURNISSEURS 
WHERE FAX IS NULL AND (PAYS = 'Allemagne' OR PAYS = 'France')

## NOT

In [ ]:
%%sql 
SELECT NOM, FONCTION, SALAIRE 
FROM EMPLOYES 
WHERE SALAIRE > 2500 AND NOT FONCTION LIKE 'Rep%'

## NOT IN

In [ ]:
%%sql 
SELECT NOM,PRENOM, TITRE  
FROM EMPLOYES 
WHERE TITRE NOT IN ('M.','Mme')

## NOT BETWEEN

In [ ]:
%%sql 
SELECT NOM, PRENOM, DATE_EMBAUCHE 
FROM EMPLOYES 
WHERE DATE_EMBAUCHE NOT BETWEEN '01/01/1990' AND '31/07/2003'
ORDER BY DATE_EMBAUCHE

## IS NOT NULL

In [ ]:
%%sql 
SELECT NOM, PRENOM, SALAIRE, COMMISSION 
FROM EMPLOYES 
WHERE COMMISSION IS NOT NULL AND SALAIRE > 10000

## NOT LIKE

In [ ]:
%%sql 
SELECT NOM_CATEGORIE, DESCRIPTION 
FROM CATEGORIES 
WHERE DESCRIPTION NOT LIKE '%' || NOM_CATEGORIE ||'%'

# Atelier 3.2	Les opérateurs logiques

## L’opérateur BETWEEN
Écrivez les requêtes permettant d’afficher :
>> * Le nom, prénom, fonction et salaire des employés qui ont un salaire compris entre **3500** et **6000**.<br>
>> * Le numéro de commande, code client et la date de commande pour les commandes passées entre le **'01/01/2011'** et **'03/01/2011'**.<br>

## La comparaison avec des listes
Écrivez les requêtes permettant d’afficher :
>> * Le nom de la société, l’adresse, le téléphone et la ville des clients qui habitent à **Toulouse**, à **Strasbourg**, à **Nantes** ou à **Marseille**.<br>
>> * Le nom du produit, le fournisseur, la catégorie et les quantités en stock pour les produits qui sont d’une des catégories 1, 3, 5 et 7.<br>
>> * Le numéro de commande, code client et la date de commande pour les commandes passées dans une des dates : **'18/02/2011'**, **'20/02/2011'** ou **'25/02/2011'**.<br>

## L’assemblage des expressions

>> * Le nom, prénom, fonction et le salaire des représentants qui sont en activité depuis **'10/10/2002'**.<br>
>> * Le nom, prénom, fonction et le salaire des employés qui sont âgés de plus de 45 ans ou qui ont une ancienneté de plus de 10 ans.<br>
>> * Le nom du produit, le fournisseur, la catégorie et les quantités des produits qui ont le numéro fournisseur entre 1 et 3 ou un code catégorie entre 1 et 3 et pour lesquelles les quantités sont données en boîtes ou en cartons.<br>
>> * Les produits et leur quantité à condition que leur emballage ne soit pas d’un de ces types : cartons, boîtes ou unités et qu’il ne soit pas conditionné par paquets de 24 ou 32. Il ne faut pas afficher les produits de catégorie 1, 4 et 8.<br>